### PR TITLE
Add support for Python 3.11.

### LINKED ISSUE #37: Does not build on Python 3.11.0a2
### What I did:

`tox -epy311`

### What I expect to happen:

Successful build and test run.

### What actually happened:

```
$ tox -epy311
py311 create: /.../ExtensionClass/.tox/py311
py311 develop-inst: /.../ExtensionClass
ERROR: invocation failed (exit code 1), logfile: /.../ExtensionClass/.tox/py311/log/py311-1.log
================================================================================================= log start ==================================================================================================
Looking in indexes: https://pypi.org/simple, https://mihowitz:****@development.verdi.de/devpi/verdi/prod/+simple/
Obtaining file:///.../ExtensionClass
Collecting zope.testrunner
  Using cached https://development.verdi.de/devpi/verdi/prod/%2Bf/ae7/fbeb862a36083/zope.testrunner-5.4.0-py2.py3-none-any.whl (216 kB)
Collecting zope.exceptions
  Using cached https://development.verdi.de/devpi/verdi/prod/%2Bf/bb9/8cc07e90ebe59/zope.exceptions-4.4-py2.py3-none-any.whl (18 kB)
Requirement already satisfied: setuptools in ./.tox/py311/lib/python3.11/site-packages (from zope.testrunner->ExtensionClass==4.5.2.dev0) (58.1.0)
Collecting zope.interface
  Using cached https://development.verdi.de/devpi/root/pypi/%2Bf/5db/a5f530fec3f09/zope.interface-5.4.0.tar.gz (249 kB)
Collecting six
  Using cached https://development.verdi.de/devpi/root/pypi/%2Bf/8ab/b2f1d86890a2d/six-1.16.0-py2.py3-none-any.whl (11 kB)
Using legacy 'setup.py install' for zope.interface, since package 'wheel' is not installed.
Installing collected packages: zope.interface, zope.exceptions, six, zope.testrunner, ExtensionClass
    Running setup.py install for zope.interface: started
    Running setup.py install for zope.interface: finished with status 'done'
  Running setup.py develop for ExtensionClass
    ERROR: Command errored out with exit status 1:
     command: /.../ExtensionClass/.tox/py311/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/.../ExtensionClass/setup.py'"'"'; __file__='"'"'/.../ExtensionClass/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
         cwd: /.../ExtensionClass/
    Complete output (33 lines):
    running develop
    running egg_info
    writing src/ExtensionClass.egg-info/PKG-INFO
    writing dependency_links to src/ExtensionClass.egg-info/dependency_links.txt
    writing requirements to src/ExtensionClass.egg-info/requires.txt
    writing top-level names to src/ExtensionClass.egg-info/top_level.txt
    reading manifest file 'src/ExtensionClass.egg-info/SOURCES.txt'
    reading manifest template 'MANIFEST.in'
    adding license file 'LICENSE.txt'
    writing manifest file 'src/ExtensionClass.egg-info/SOURCES.txt'
    running build_ext
    building 'ExtensionClass._ExtensionClass' extension
    creating build/temp.macosx-11.0-x86_64-3.11
    creating build/temp.macosx-11.0-x86_64-3.11/src
    creating build/temp.macosx-11.0-x86_64-3.11/src/ExtensionClass
    /usr/bin/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -pipe -Os -isysroot/Library/Developer/CommandLineTools/SDKs/MacOSX11.sdk -Isrc -I/.../ExtensionClass/.tox/py311/include -I/opt/local/Library/Frameworks/Python.framework/Versions/3.11/include/python3.11 -c src/ExtensionClass/_ExtensionClass.c -o build/temp.macosx-11.0-x86_64-3.11/src/ExtensionClass/_ExtensionClass.o
    src/ExtensionClass/_ExtensionClass.c:840:16: error: expression is not assignable
      Py_TYPE(typ) = ECExtensionClassType;
      ~~~~~~~~~~~~ ^
    src/ExtensionClass/_ExtensionClass.c:800:33: warning: comparison of integers of different signs: 'Py_ssize_t' (aka 'long') and 'unsigned long' [-Wsign-compare]
              if (typ->tp_basicsize <= sizeof(_emptyobject))
                  ~~~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~~~~~
    src/ExtensionClass/_ExtensionClass.c:985:32: error: expression is not assignable
      Py_TYPE(&ExtensionClassType) = &PyType_Type;
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
    src/ExtensionClass/_ExtensionClass.c:995:22: error: expression is not assignable
      Py_TYPE(&BaseType) = &ExtensionClassType;
      ~~~~~~~~~~~~~~~~~~ ^
    src/ExtensionClass/_ExtensionClass.c:1003:42: error: expression is not assignable
      Py_TYPE(&NoInstanceDictionaryBaseType) = &ExtensionClassType;
      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^
    1 warning and 4 errors generated.
    error: command '/usr/bin/clang' failed with exit code 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /.../ExtensionClass/.tox/py311/bin/python -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/.../ExtensionClass/setup.py'"'"'; __file__='"'"'/.../ExtensionClass/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.
```

### What version of Python and Zope/Addons I am using:

MacOS 11.6.1
Python 3.11.0a2

### PR DESCRIPTION
**Pull request by icemac**
Fixes #37.

I ran https://github.com/pythoncapi/pythoncapi_compat on the code to make it compatible with Python 3.11.

Problems
========

- [x] `pythoncapi_compat.h` is required so the code stays compatible with Python < 3.10. This file is MIT licensed and comes from a non-Zope contributor. So the license agreement forbids to commit it to zopefoundation repositories. – Any idea how to work around this issue?
- [x] The C version for Python 3.11 has a problem that sub-classes of `ExtensionClass.NoInstanceDictionaryBase` should also not have a `__dict__` but the tests shows they have it. – I see these failures locally on MacOS with `alpha3`.
- [x] The change does not seem to be compatible with Windows.

### CODE REVIEW

File: .manylinux-install.sh [hunk 1]
```diff
@@ -23,15 +23,20 @@ fi
...
+       [[ "${PYBIN}" == *"cp27"* ]] || \
        [[ "${PYBIN}" == *"cp35"* ]] || \
+       [[ "${PYBIN}" == *"cp311"* ]] || \
        [[ "${PYBIN}" == *"cp36"* ]] || \
        [[ "${PYBIN}" == *"cp37"* ]] || \
        [[ "${PYBIN}" == *"cp38"* ]] || \
-       [[ "${PYBIN}" == *"cp310"* ]] || \
-       [[ "${PYBIN}" == *"cp39"* ]]; then
+       [[ "${PYBIN}" == *"cp39"* ]] || \
+       [[ "${PYBIN}" == *"cp310"* ]] ; then
```
Comment (d-maurer) [inline .manylinux-install.sh:39]: Why no "cp311"?

**Review by d-maurer — APPROVED** [review]

**Comment by d-maurer** [comment]
* `pythoncapi_compat.h`: likely, we need only a few elementary definitions (e.g. `_Py_SET_TYPE`). We could create them ourselves.
* Windows: apparently, the compiler does not like something in `pythoncapi_compat.h`. The problem might disappear when we use only the really required definitions.
* The code tries to inhibit the instance dict creation by setting `tp_dictoffset` to `0`. I have no idea why this does not work (the doc still suggests that it should).

**Comment by dataflake** [comment]
I agree with Dieter, we could just copy the few bits we need from `pythoncapi_compat.h`. I don't see another workaround. I would not feel comfortable taking the file as a whole due to licensing issues, and I would guess the developers won't re-license it for us.

File: .manylinux-install.sh (hunk 1 above)
Comment (icemac) [reply .manylinux-install.sh:39]: It gets added in line 34. The Python versions are not in order to make it easier to add specific ones via configuration, see https://github.com/zopefoundation/meta/blob/fba6d957ba447b6fa369d872e803756bd5176391/config/c-code/manylinux-install.sh.j2#L28-L41 for the code which produces this file.

**Comment by icemac** [comment]
Removing the lines the Windows C compiler complains about seems to fix this issue for Windows as they are not needed for ExtensionClass.

**Comment by icemac** [comment]
@d-maurer wrote:
> * The code tries to inhibit the instance dict creation by setting `tp_dictoffset` to `0`. I have no idea why this does not work (the doc still suggests that it should).

It could be a bug in Python 3.11.0a3 which I am using locally. GitHub actions is still on a2 where this problem does not occur. Let's see what happens when a3 is available there.

**Comment by d-maurer** [comment]
Michael Howitz wrote at 2021-12-20 23:25 -0800:
***@***.*** wrote:
>> * The code tries to inhibit the instance dict creation by setting `tp_dictoffset` to `0`. I have no idea why this does not work (the doc still suggests that it should).
>
>It could be a bug in Python 3.11.0a3 which I am using locally. GitHub actions is still on a2 where this problem does not occur. Let's see what happens when a3 is available there.

I am analysing this issue.

Apparently, `type_new` has already created a `__dict__` slot
before `EC_new` could set `tp_dictoffset` to `0`.


For my local examination, I use the following `compat.h`
```
#include <Python.h>

#ifndef _PyObject_CAST
#  define _PyObject_CAST(op) ((PyObject*)(op))
#endif

#if PY_VERSION_HEX < 0x030900A4 && !defined(Py_SET_TYPE)
static inline void
_Py_SET_TYPE(PyObject *ob, PyTypeObject *type)
{
    ob->ob_type = type;
}
#define Py_SET_TYPE(ob, type) _Py_SET_TYPE(_PyObject_CAST(ob), type)
#endif
```

**Comment by d-maurer** [comment]
Michael Howitz wrote at 2021-12-20 23:25 -0800:
***@***.*** wrote:
>> * The code tries to inhibit the instance dict creation by setting `tp_dictoffset` to `0`. I have no idea why this does not work (the doc still suggests that it should).
>
>It could be a bug in Python 3.11.0a3 which I am using locally. GitHub actions is still on a2 where this problem does not occur. Let's see what happens when a3 is available there.

3.11a3 has introduced a new type flag `Py_TPFLAGS_MANAGED_DICT`.
If this flag is set, `tp_dictoffset` has no effect.

For heap allocated type objects, `Py_TPFlags_MANAGED_DICT` is
often set automatically, among then for `NoInstanceDictionaryBase`
subtypes.

The following patch lets the test suite succeed for 3.11a3:
```
--- _ExtensionClass.c~	2021-12-21 08:59:20.468180694 +0100
+++ _ExtensionClass.c	2021-12-22 07:53:14.618794977 +0100
@@ -338,6 +338,9 @@
               )
             {
               TYPE(result)->tp_dictoffset = 0;
+#ifdef Py_TPFLAGS_MANAGED_DICT
+	      TYPE(result)->tp_flags &= ~Py_TPFLAGS_MANAGED_DICT;
+#endif
               break;
             }
         }
```

**Comment by vstinner** [comment]
> pythoncapi_compat.h is required so the code stays compatible with Python < 3.10. This file is MIT licensed and comes from a non-Zope contributor. So the license agreement forbids to commit it to zopefoundation repositories. – Any idea how to work around this issue?

Hi, I'm the author of this header file. I don't understand the license issue, would you mind to elaborate?

**Comment by dataflake** [comment]
Hi @vstinner, the old Zope Foundation and the Plone Foundation that absorbed it hold the copyright for all code in the zopefoundation GitHub repositories. All contributors must sign a contributor agreement that co-assigns copyrights to the Plone Foundation - something which only the original code author can legally do. Copying in third-party code breaks the contributor agreement and potentially leaves the Plone Foundation and/or the contibutor exposed to copyright claims. Also, code within the zopefoundation GitHub repositories is usually licensed under the Zope Public License (ZPL) to minimize hassle for end users that worry about licensing issues. A mix of licenses or no license at all is a turn-off for e.g. businesses.

The "normal" process for contributors involves them signing the contributor agreement and then only contributing code they have authored themselves.

**Comment by vstinner** [comment]
I'm open to add a second license to the pythoncapi_compat.h header, but I don't know what to do with the copyright part. Also, it's kind of weird that I would have to add a new license, transfer the copyright, and/or sign a contributor agreement, each time a project wants to use my header file. I'm always lost when it comes to license and copyright. I'm French, the law is different than in the USA.

**Comment by dataflake** [comment]
@vstinner In this case the safest approach for us is to just recreate the very few parts we need from `pythoncapi_compat.h` ourselves and not use your file.

**Comment by vstinner** [comment]
> @vstinner In this case the safest approach for us is to just recreate the very few parts we need from pythoncapi_compat.h ourselves and not use your file.

Sure. That makes sense.

**Comment by icemac** [comment]
I updated the PR to run on Python 3.11a3 on GHA. Is there anything left to do before this PR can be merged?

**Comment by dataflake** [comment]
I think it's OK. Tests run fine.

**Comment by icemac** [comment]
Released to PyPI: https://pypi.org/project/ExtensionClass/4.6/

**Comment by icemac** [comment]
Thank you for working on this PR and fixing the issues. 😃

**Pull request merged**